src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -0,0 +1,253 @@
+				(SymbolKind.Field, SymbolKind.Method) => targetsType ?
+					DiagnosticId.DynamicallyAccessedMembersMismatchFieldTargetsMethodReturnType :
+					DiagnosticId.DynamicallyAccessedMembersMismatchFieldTargetsMethod,
+				(SymbolKind.Field, SymbolKind.TypeParameter) => DiagnosticId.DynamicallyAccessedMembersMismatchFieldTargetsGenericParameter,
+				(SymbolKind.NamedType, SymbolKind.Method) => targetsType ?
+					DiagnosticId.DynamicallyAccessedMembersMismatchMethodTargetsMethodReturnType :
+					DiagnosticId.DynamicallyAccessedMembersMismatchMethodTargetsMethod,
+				(SymbolKind.Method, SymbolKind.Field) => DiagnosticId.DynamicallyAccessedMembersMismatchMethodReturnTypeTargetsField,
+				(SymbolKind.Method, SymbolKind.Method) => targetsType ?
+					DiagnosticId.DynamicallyAccessedMembersMismatchMethodReturnTypeTargetsMethodReturnType :
+					DiagnosticId.DynamicallyAccessedMembersMismatchMethodReturnTypeTargetsMethod,
+				(SymbolKind.Method, SymbolKind.Parameter) => targetsType ?
+					DiagnosticId.DynamicallyAccessedMembersMismatchMethodTargetsParameter :
+					DiagnosticId.DynamicallyAccessedMembersMismatchMethodReturnTypeTargetsParameter,
+				(SymbolKind.NamedType, SymbolKind.Field) => DiagnosticId.DynamicallyAccessedMembersMismatchMethodTargetsField,
+				(SymbolKind.TypeParameter, SymbolKind.Field) => DiagnosticId.DynamicallyAccessedMembersMismatchTypeArgumentTargetsField,
+				(SymbolKind.TypeParameter, SymbolKind.Method) => DiagnosticId.DynamicallyAccessedMembersMismatchTypeArgumentTargetsMethodReturnType,
+				(SymbolKind.TypeParameter, SymbolKind.Parameter) => DiagnosticId.DynamicallyAccessedMembersMismatchTypeArgumentTargetsParameter,
+				(SymbolKind.TypeParameter, SymbolKind.TypeParameter) => DiagnosticId.DynamicallyAccessedMembersMismatchTypeArgumentTargetsGenericParameter,
+				_ => throw new NotImplementedException ()
+			};
+
+		static string[] GetDiagnosticArguments (ISymbol source, ISymbol target, string missingAnnotations)
+		{
+			var args = new List<string> ();
+			args.AddRange (GetDiagnosticArguments (target));
+			args.AddRange (GetDiagnosticArguments (source));
+			args.Add (missingAnnotations);
+			return args.ToArray ();
+		}
+
+		static IEnumerable<string> GetDiagnosticArguments (ISymbol symbol)
+		{
+			var args = new List<string> ();
+			args.AddRange (symbol.Kind switch {
+				SymbolKind.Parameter => new string[] { symbol.GetDisplayName (), symbol.ContainingSymbol.GetDisplayName () },
+				SymbolKind.NamedType => new string[] { symbol.GetDisplayName () },
+				SymbolKind.Field => new string[] { symbol.GetDisplayName () },
+				SymbolKind.Method => new string[] { symbol.GetDisplayName () },
+				SymbolKind.TypeParameter => new string[] { symbol.GetDisplayName (), symbol.ContainingSymbol.GetDisplayName () },
+				_ => throw new NotImplementedException ($"Unsupported source or target symbol {symbol}.")
+			});
+
+			return args;
+		}
+
+		static ISymbol? GetSymbolFromOperation (IOperation? operation) =>
```

Comment:
Maybe `TryGet` to signal that it might return `null`?

In general, this is an impossible function to write because IOperations can represent 0 or more symbols, so possibly failing seems like the right semantic.